hexl/thread-pool/thread-pool.hpp

```diff
@@ -0,0 +1,316 @@
+        }
+        thread_handler->thread.join();
+
+        // Update handlers
+        delete thread_handler;
+        auto it = thread_handlers.end() - 1;
+        thread_handlers.erase(it);
+        total_threads--;
+      }
+    }
+
+    if (n_threads == 0) {
+      setup_done = false;
+    }
+
+    HEXL_VLOG(2,
+              "Setting up thread pool with " << total_threads << " threads.");
+  }
+
+ public:
+  // Methods
+
+  ThreadPool() { ThreadPool::child = false; }
+
+  ~ThreadPool() { SetupThreads(0); }
+
+  // GetNumThreads: Returns total number of threads
+  size_t GetNumThreads() {
+    std::lock_guard<std::mutex> lock(pool_mutex);
+    return total_threads;
+  }
+
+  // AddParallelJobs: Runs the same function on a total number of threads
+  void AddParallelJobs(tp_task_t job) {
+    HEXL_CHECK(job, "Require non empty job");
+
+    // Try using thread pool
+    if (pool_mutex.try_lock()) {
+      if (!setup_done) {
+        SetupThreads_Unlocked(HEXL_NUM_THREADS);  // Setup if thread pool down
+      }
+
+      // Only if all threads are available
+      if (next_thread.load() == 0) {
+        next_thread.store(total_threads);
+
+        for (size_t i = 0; i < total_threads; i++) {
+          thread_info_t* thread_handler = thread_handlers.at(i);
+          if (thread_handler->state.load() == static_cast<int>(STATE::DONE)) {
+            thread_handler->task = job;
+            thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+          } else if (thread_handler->state.load() ==
+                     static_cast<int>(STATE::SLEEPING)) {
+            thread_handler->task = job;
+            thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+            thread_handler->waker.notify_one();
+          } else {  // In case thread is not on expected state
+            job(i, total_threads);
+          }
+        }
+        SetBarrier_Unlocked();  // Wait 'til all jobs are done
+        pool_mutex.unlock();
+      } else {  // Run on single thread
+        pool_mutex.unlock();
+        job(0, 1);
+      }
+    } else {
+      job(0, 1);
+    }
+  }
+
+  // AddRecursiveCalls: Runs a task on next thread available
+  void AddRecursiveCalls(tp_task_t task_a, tp_task_t task_b) {
+    HEXL_CHECK(task_a, "task_a: Require non empty task");
+    HEXL_CHECK(task_b, "task_b: Require non empty task");
+    bool locked = false;
+
+    // Try using thread pool
+    if (!ThreadPool::child) {
+      locked = pool_mutex.try_lock();
+      if (!locked) {
+        task_a(0, 1);
+        task_b(0, 1);
+        return;
+      }
+    }
+
+    if (!setup_done) {
+      SetupThreads_Unlocked(HEXL_NUM_THREADS);  // Setup if thread pool down
+    }
+
+    size_t next = next_thread.fetch_add(2);
+    if (next <= total_threads - 2) {
+      thread_info_t* thread_handler = thread_handlers.at(next++);
+      if (thread_handler->state.load() == static_cast<int>(STATE::DONE)) {
```

Comment:
since it is all == on STATE I think you should use switch case to clear up the code